### ISSUE TITLE
Flutter Web: Image in Container makes shadow different.

### ISSUE THREAD

**Issue by atfa** [flutter/flutter#55933]
## Steps to Reproduce

1. Enter code below and put a 200x200 logo.jpg file into assets folder.
```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _showPic = 1;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(title: Text(widget.title)),
        body: Padding(
            padding: EdgeInsets.fromLTRB(0, 50, 0, 0),
            child: Center(
                child: Column(children: [
              Container(
                width: 300,
                height: 300,
                decoration: BoxDecoration(
                    boxShadow: [BoxShadow(color: Colors.black26, blurRadius: 26.0)],
                    shape: BoxShape.rectangle,
                    borderRadius: BorderRadius.all(Radius.circular(20.0)),
                    color: Colors.white),
                child: _showPic % 2 == 0 ? Icon(Icons.event) : Image.asset("logo.jpg")
              ),
              Padding(
                  padding: EdgeInsets.fromLTRB(0, 50, 0, 0),
                  child: RaisedButton(
                      child: Text("Switch"), onPressed: () => setState(() => _showPic++)))
            ]))) // This trailing comma makes auto-formatting nicer for build methods.
        );
  }
}
```
2. Run `flutter build web`.
3. Server web using any http server

**Expected results:** 
The shadow effect on both Android+Chrome and iOS+Safari should be the same.

**Actual results:** 
If there is a **Image** in the Container, the shadow looks different on both  Android+Chrome and iOS+Safari.

Below is the web app under Android+Chrome. There is a **Icon** in the Container.
![android-chrome-icon](https://user-images.githubusercontent.com/16986056/80558829-3d3ffb80-8a0e-11ea-9ef6-6bc0ea808cb7.png)

Below is the web app under Android+Chrome. There is a **Image** in the Container. The shadow seems lighter and thiner.
![android-chrome-img](https://user-images.githubusercontent.com/16986056/80558761-07027c00-8a0e-11ea-9a6b-69a1b6344421.png)

Below is the web app under iOS+Safari. There is a **Icon** in the Container.
![Simulator-iPhoneX-icon](https://user-images.githubusercontent.com/16986056/80558944-a293ec80-8a0e-11ea-8155-dc4e42a4b949.png)

Below is the web app under iOS+Safari. There is a **Image** in the Container. The shadow disappeared.
![Simulator-iPhoneX-img](https://user-images.githubusercontent.com/16986056/80558935-9576fd80-8a0e-11ea-816d-e955658f7337.png)

## flutter doctor -v

```
$ flutter doctor -v
[✓] Flutter (Channel dev, v1.18.0-8.0.pre, on Mac OS X 10.13.6 17G2208, locale en-CN)
    • Flutter version 1.18.0-8.0.pre at /Users/atfa/flutter
    • Framework revision e0c63cd35e (5 days ago), 2020-04-23 22:29:01 -0400
    • Engine revision d2ec21221e
    • Dart version 2.9.0 (build 2.9.0-3.0.dev b0d35855d8)

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/atfa/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    ✗ Android license status unknown.
      Try re-installing or updating your Android SDK Manager.
      See https://developer.android.com/studio/#downloads or visit visit
      https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions.

[!] Xcode - develop for iOS and macOS (Xcode 9.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.3, Build version 9E145
    ✗ Flutter requires a minimum Xcode version of 11.0.0.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.44.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.9.1

[!] Proxy Configuration
    • HTTP_PROXY is set
    ! NO_PROXY is not set

[✓] Connected device (3 available)
    • macOS      • macOS      • darwin-x64     • Mac OS X 10.13.6 17G2208
    • Web Server • web-server • web-javascript • Flutter Tools
    • Chrome     • chrome     • web-javascript • Google Chrome 81.0.4044.122

! Doctor found issues in 3 categories.
```

</details>

**Comment by VladyslavBondarenko** [comment]
Link to reproduce https://flwt.netlify.app/

Additional note about safari (for both mobile/desktop browser):

On the first loading:
![image](https://user-images.githubusercontent.com/17024977/80575001-82aef980-8a0b-11ea-93b2-f4d7cd9c6906.png)

After switching to icon and back (you can see half-transparent corners):
![image](https://user-images.githubusercontent.com/17024977/80575065-98242380-8a0b-11ea-88aa-452a11dbad69.png)

**Comment by yjbanov** [comment]
I've looked into it a bit and the behaviors described here seem to be caused by 3 separate issues:

- Safari does not support the `filter` property on 2D canvas (this is being fixed in https://github.com/flutter/engine/pull/18216)
- The shadow is painted using two methods depending on whether you use an icon vs an image. When using an icon the engine decides to render via the `DomCanvas`, and when using an image it decides to render using `BitmapCanvas`. We need to make shadows consistent between the two.
- There seems to be a bug in Safari where leaves a phantom box with a shadow after we have removed it (simply disabling `visibility` on the parent `<flt-clip>` and enabling it again in Web Inspector fixes the issue). I don't have a solution for this yet. Ideas welcome!

**Comment by yjbanov** [comment]
Update:

> * Safari does not support the `filter` property on 2D canvas (this is being fixed in [flutter/engine#18216](https://github.com/flutter/engine/pull/18216))

This has landed, and I checked that it fixes the case when no shadow is visible at all.

> * The shadow is painted using two methods depending on whether you use an icon vs an image. When using an icon the engine decides to render via the `DomCanvas`, and when using an image it decides to render using `BitmapCanvas`. We need to make shadows consistent between the two.

The issue here was that the amount of `MaskFilter` blur was not adjusted by device-pixel ratio. The CSS side of the canvas is device-pixel ratio bigger than its bitmap pixel size (we do this to avoid pixelation). I am working on a fix for this now.

> * There seems to be a bug in Safari where it leaves a phantom box with a shadow after we have removed it (simply disabling `visibility` on the parent `<flt-clip>` and enabling it again in Web Inspector fixes the issue). I don't have a solution for this yet. Ideas welcome!

No update yet.

**Comment by yjbanov** [comment]
I split out the Safari DOM update bug into a standalone github issue: https://github.com/flutter/flutter/issues/57527
